File: data-structures/java/classwork/treeSearches/BFS.java
// Class to represent Tree node
class Node {
    int data;
    Node left, right;

    public Node(int item) {
        data = item;
        left = null;
        right = null;
    }
}

public class BFS {
    // Root of the Binary Tree
    Node root;

    public BFS() {
        root = null;
    }

    // find the "height" of a tree (the number of nodes along the
    // longest path from the root node down to the farthest leaf node.)
    public int height(Node root) {
        // empty tree has height 0
        if (root == null)
            return 0;
        else {
            // compute height of each subtree
            int lheight = height(root.left);
            int rheight = height(root.right);

            // return the larger of the two heights
            if (lheight > rheight)
                return (lheight + 1);
            else
                return (rheight + 1);
        }
    } // height

    // print level order traversal of tree
    public void printLevelOrder() {
        int h = height(root);
        for (int i = 1; i <= h; i++)
            printCurrentLevel(root, i);
    } // printLevelOrder

    // print nodes at the current level
    public void printCurrentLevel(Node root, int level) {
        if (root == null)
            return;
        // print out the first level
        if (level == 1)
            System.out.print(root.data + " ");
        // recursively print out the following levels
        else if (level > 1) {
            printCurrentLevel(root.left, level - 1);
            printCurrentLevel(root.right, level - 1);
        }
    } // printCurrentLevel

    public static void main(String args[]) {
        BFS tree = new BFS();
        tree.root = new Node(4);
        tree.root.left = new Node(2);
        tree.root.right = new Node(6);
        tree.root.left.left = new Node(1);
        tree.root.left.right = new Node(3);
        tree.root.right.left = new Node(5);
        tree.root.right.right = new Node(7);

        System.out.println("Level order traversal of the binary tree is ");
        tree.printLevelOrder();
    } // main
}

File: data-structures/python/classwork/treeSearches/BFS.py
class Node:
    def __init__(self, data):
        self.left = None
        self.right = None
        self.data = data


def height(node):
    """find the "height" of a tree (the number of nodes along the
    longest path from the root node down to the farthest leaf node.)
    """
    # empty tree has height 0
    if node is None:
        return 0
    else:
        # compute the height of each subtree
        lheight = height(node.left)
        rheight = height(node.right)

        # return the larger of the two heights
        if lheight > rheight:
            return lheight + 1
        else:
            return rheight + 1


# print nodes at a current level
def print_current_level(root, level):
    if root is None:
        return
    # print out the first level
    if level == 1:
        print(root.data, end=" ")
    # recursively print out the following levels
    elif level > 1:
        print_current_level(root.left, level - 1)
        print_current_level(root.right, level - 1)


# print level order traversal of tree
def print_level_order(root):
    h = height(root)
    for i in range(1, h + 1):
        print_current_level(root, i)


if __name__ == "__main__":

    root = Node(4)
    root.left = Node(2)
    root.right = Node(6)
    root.left.left = Node(1)
    root.left.right = Node(3)
    root.right.left = Node(5)
    root.right.right = Node(7)

    print("Level order traversal of the binary tree is")
    print_level_order(root)
